### ISSUE TITLE
Rubocop brakes on 2.7 pattern matching AST nodes while tree walking, as they children may be `nil`

### ISSUE THREAD

**Issue by maksar** [rubocop-hq/rubocop#7624]
Following code
```ruby
  case payload
  in key:
    method
  end
```
produces such AST (with `Parser::CurrentRuby.parse`, using `Parser::Ruby27` parser gem version: `2.7.0.1`):

```ruby
s(:case_match,
  s(:send, nil, :payload),
  s(:in_pattern,
    s(:hash_pattern,
      s(:match_var, :key)), nil, # <- here
    s(:send, nil, :method)), nil)
```

Second children of the `in_pattern` s-expression is `nil`, which leads to 
```
`undefined method `type' for nil:NilClass
```
error on line https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/ast/traversal.rb#L57

Simplest fix (https://github.com/maksar/rubocop/commit/db9f1cff43cdcbb25318bc13e0eb10ee2e3287e0) for me was adding `.compact`, but I'm not sure it is the right thing to do.

Issue is reproducible with current master (019fe5e3f0bdd0a533e51acc0c864330416cf08b)

**Comment by buehmann** [comment]
Thanks! This is related to my remark here: https://github.com/rubocop-hq/rubocop/pull/7605#discussion_r361961642

**Comment by iliabylich** [comment]
Why does rubocop have its own visitor implementation? parser ships with https://github.com/whitequark/parser/blob/master/lib/parser/ast/processor.rb

**Comment by buehmann** [comment]
According to Git history:
```
Add RuboCop::Node::Traversal (to traverse AST without mutating it)

When AST::Processor calls a hook, it takes the return value as an updated
node and tries building a modified AST. We don't want to do this.
```
https://github.com/rubocop-hq/rubocop/commit/aead3449c3f32a9386fc69b32031d76addf42d22

**Comment by koic** [comment]
This issue will solved by #7786.
